streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala

```diff
@@ -2868,6 +2868,33 @@ object ZStreamSpec extends ZIOBaseSpec {
               _     <- f.join
             } yield assertTrue(count == 0)
           } @@ TestAspect.jvmOnly @@ nonFlaky,
+          test("parallelism is controlled by Ref") {
+            // each second parallelism is updated
+            // then counts how many increments have been done each second
+            val iterations = 200
+            for {
+              parallelism <- Ref.make(1)
+              counter     <- Ref.make(Map[Long, Int]())
+              f0 <- ZStream
+                      .range(0, iterations)
+                      .mapZIOPar(parallelism, bufferSize = 32)(_ =>
+                        (for {
+                          second <- Clock.instant.map(_.getEpochSecond)
+                          _      <- counter.update(_.updatedWith(second)(_.map(_ + 1).orElse(Some(1))))
+                        } yield ()).delay(99.millis)
+                      )
+                      .runDrain
+                      .fork
+
+              f1      <- parallelism.set(2).delay(1000.millis).fork
+              f2      <- parallelism.set(4).delay(2000.millis).fork
+              _       <- ZIO.iterate(0)(_ <= iterations + 3)(i => TestClock.adjust(100.millis).as(i + 1))
+              _       <- f0.join
+              _       <- f1.join
+              _       <- f2.join
+              buckets <- counter.get.map(_.toList.sortBy(_._1).take(3).map(_._2))
+            } yield assertTrue(buckets == List(10, 20, 39)) // off by one
+          } @@ TestAspect.jvmOnly @@ nonFlaky(3),
```

Comment:
this tests seems to work, but it is slow vs the rest, and there's the off by one. looking for other ideas how to test this

Reply:
improved here the test time